### ISSUE TITLE
Question: working principle of RecursiveCheckpointAdjoint

### ISSUE THREAD

**Issue by TommyGiak** [patrick-kidger/diffrax#564]
Hi! I’ve recently started diving deeper into Neural ODEs, but I feel like I’m missing some foundational concepts regarding checkpointing in backpropagation.

Could you help clarify how backpropagation is implemented through checkpointing in RecursiveCheckpointAdjoint? Specifically, I have a few questions:
1. What exactly is saved during the forward pass[^1]?
1. Is the gradient (with respect to the loss) computed purely through the internal operations of the solver, after recomputing the forward pass using the saved states?
1. If we need to recompute all the states of the “reverse equation,” why does using more checkpoints result in faster computation?
1. Why does the implementation need to be recursive to save memory during backpropagation? Could a non-recursive implementation achieve the same efficiency?

I’m a bit confused about these concepts right now 🫠. Thanks in advance for your help!

[^1]: From what I understand, only the states of the ODE solve at certain points (determined by the number of checkpoints) are saved, but what about the computational graph?

**Comment by lockwo** [comment]
I'll just remark on a few things I know that might be helpful.

1. What's being saved? Basically the state at that point, you can see in this example: 

```python
import diffrax
import jax
from jax import numpy as jnp
from jax.ad_checkpoint import print_saved_residuals

vector_field = lambda t, y, args: -y
term = diffrax.ODETerm(vector_field)
solver = diffrax.Euler()
saveat = diffrax.SaveAt(ts=jnp.linspace(0, 1, 10))

def f(y):
  return diffrax.diffeqsolve(term, solver, t0=0, t1=1, dt0=0.01, y0=y, saveat=saveat, adjoint=diffrax.RecursiveCheckpointAdjoint(4000)).ys.mean()

print(jax.ad_checkpoint.print_saved_residuals(f, jnp.array(1.0)))
```

The output has something like 

```
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
bool[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
bool[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
f32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
f32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
f32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
bool[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
f32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
i32[4000] output of jitted function 'diffeqsolve' from /usr/local/lib/python3.10/dist-packages/equinox/_jit.py:244 (_call)
```

(because the state is a bunch of arrays). If you change this to be 1 checkpoint, you will see these arrays are size 1. 


2. I'm not sure I understand this one, what else would the gradient depend on?

3. As hinted above, the adjoint has a nice tradeoff between memory and speed. If you have checkpointing every step you store more intermediate results for your gradient computation (as you can see the arrays are larger), thus you're speed is just the standard cost of backprop. However, if you checkpoint less often, you store less memory (but you have to recompute things in order to actually get the information you need for backprop, thus it increases time). The reason this matters is because if you are solving a DE, you might have a much higher depth than something like in traditional ML (e.g. I have set adaptive solvers with max_steps of 100 million, which is not always a tractable array size to store in memory).

4. I don't know as much on this one. All I'll say is I believe the "recursive" applies to the sort of multi-leveled approach of while loops both being checkpoints (e.g. the fact that there are two while loops, one for the integration over time and one for the interpolation/saving potentially between solver steps, so you're like "recursing" levels of gradient computation, see https://cseweb.ucsd.edu/~tzli/cse291/sp2024/lectures/checkpointing.pdf for more).

**Comment by patrick-kidger** [comment]
This is a great topic.

1. What is being saved? An ODE solver works by iteratively taking many steps, `y_{n+1} = F(y_n)`. For example Euler's method has `F(y_n) = y_n + f(y_n)dt`. We pick some values of `n` for which we save `y_n`. We do not save the other `y_n`, nor do we save the internal computations of `F`.

    The number of `n` that we are willing to save is our number of checkpoints. You can see this in @lockwo's example above, with `RecursiveCheckpointAdjoint(<number of checkpoints>)`.

2. Yes, the gradient is computed by working backwards through the internal operations solver. (There are some variant adjoint methods out there like `BacksolveAdjoint` that do other things, but they're mostly a terrible idea and should not be used.)

3. If we don't save a value during the forward pass, then we must recompute it during the backward pass. More checkpoints => more values saved => less recomputation needed => faster.

4. The recursive-ness is very important! Given an ODE solver that takes `n` steps, the optimal non-recursive algorithm takes `O(n^1.5)` amount of time and `O(n^0.5)` amount of memory. Meanwhile the optimal recursive algorithm takes `O(n log n)` time and `O(log n)` memory -- which is much better!

I'd recommend reading up on checkpointing following @lockwo's link above -- in particular they reference the paper on recursive checkpointing by Griewank, which is worth reading. This paper is the foundation for all kinds of recursive checkpointed methods. A little later in this link they also discuss online checkpointing by Stumm and Walther, which is an extension to handle cases where you don't know the number of steps in advance (like in an ODE solver with an adaptive number of steps, e.g. when using `diffrax.PIDController`).

Under the hood, Diffrax actually uses an extended version of the Stumm--Walther algorithm (with some elements from Wang and Moin). It's a very complicated algorithm (probably the most difficult piece of code I've ever written!) but if you're feeling masochistic then it's available here:

https://github.com/patrick-kidger/equinox/blob/main/equinox/internal/_loop/checkpointed.py

**Comment by patrick-kidger** [comment]
Right, I understand what you're thinking -- that speed would not improve when you add more checkpoints, because regardless, you're basically just always doing the original forward computation, plus an additional local forward+backward computation on the backward pass. This reasoning holds as long as you have enough memory that you're basically just not memory constrained at all! In your case: if you have `n` steps and `k` checkpoints, then you're assuming you have `O(k + n/k)` amount of memory available. (E.g. 4 checkpoints + n/4 amount of computation graph) Note that if you optimize this for the best asymptotic coefficient then you get `k=sqrt(n)` and recover the optimal rates given in my previous message.

However if you *are* memory constrained -- e.g. maybe you lay down 4 checkpoints equally, but only have enough memory to save 1/8 of the forward pass -- then in fact we would have to do our backward pass something like this:
```
x---x---x---x---
            --->
              <-
            ->
            <-
        --->
          <-
        ->
        <-
    --->
      <-
    ->
    <-
--->
  <-
->
<-
```
and here we actually have to compute our way through the first half of each piece *twice*, not just once. In such a scenario the recursive structure of the checkpointing starts to become useful.

**Comment by TommyGiak** [comment]
Yes, exactly, now it makes sense to me.
I had misunderstood the concept of memory constraint.
Thank you
